### ISSUE TITLE
The MeTTa interpreter crashes potentially leaky evaluation

### ISSUE THREAD

**Issue by ngeiswei** [trueagi-io/hyperon-experimental#313]
## Overview

The MeTTa interpreter crashes seemingly when calling a function with a variable in argument, and unification with that function's definition and is not sufficient to carry out the reduction.

## How to reproduce

Run the following metta code

```
(= (foo (A $x)) $x)
!(foo $y)
```

## What should be expected

One possible output could be `[$x#1]`, but I'm not sure if that would be permitted.

## What do we get instead

The interpreter crashes with the error message

```
thread '<unnamed>' panicked at 'called `Option::unwrap()` on a `None` value', /home/nilg/Work/TrueAGI/hyperon-experimental/lib/src/metta/interpreter.rs:573:67
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
fatal runtime error: Rust panics must be rethrown
Aborted (core dumped)
```

**Comment by vsbogd** [comment]
`(foo $y)` in the program above means we postpone the real calculations of the `foo` result until the moment when the value of the `$y` variable is known. The result of `foo` is conserved in variable bindings `{ $X, $y <- (A $X) }`. When we know the value of the `$y` then it is unified with `(A $X)` and thus the value of `$X` is determined.

**Comment by ngeiswei** [comment]
> `(foo $y)` in the program above means we postpone the real calculations of the `foo` result until the moment when the value of the `$y` variable is known.

OK, meaning that the following piece of code
```
(= (foo (A $x)) $x)
!(let* (($z (foo $y)) ($y (A a))) $z)
```
should output
```
[a]
```
Correct?

**Comment by ngeiswei** [comment]
Yes, that is what your fix returns :-)

**Comment by vsbogd** [comment]
Should be fixed by #315
